docs/DesignDocs/RemoteAttestationCollaterals.md

```diff
@@ -0,0 +1,292 @@
+Remote Attestation Collaterals
+=====
+
+Adding attestation APIs to support collaterals for remote attestation.  Collaterals are information from a second source that the user/challenger can provide for attestation verification.
+
+Motivation
+----------
+
+Currently, the existing collaterals used for Intel SGX quote verification are not exposed to the user.  Adding these new APIs allows the user to specify these collaterals instead of using the default collaterals used by OE SDK.  This enables two possible user scenarios:
+1. The user/challenger has the option to specify its set of collaterals during verification.
+2. The user/challenger can perform the verification on evidence and collaterals that were created in the past.  The user can audit the verification results anytime he/she wants.  The user will have to provide a datetime to verify against.
+
+
+User Experience
+---------------
+
+There are 3 scenarios.
+
+### 1. Challenger does not specify the collaterals:
+In this scenario the challenger does not receive any collaterals from the attester(where the TEE resides).  The challenger also does not pass any collaterals to the verification function, instead the collaterals are fetched during verification by the OE SDK.
+
+##### Attester generates the evidence (inside an enclave/TEE)
+```C
+// Get OE report
+result = oe_get_report(
+    OE_REPORT_FLAGS_REMOTE_ATTESTATION,
+    NULL, // report_data
+    0,
+    NULL, // opt_params
+    0,
+    &report,
+    report_size);
+```
+
+##### Challenger verifies the evidence (in the untrusted host)
+```C
+...
+
+// Verify report without collaterals
+result = oe_verify_remote_report_with_collaterals(
+            report,
+            report_size,
+            NULL, // collaterals
+            0,    // collaterals_size
+            NULL, // input_validation_time
+            parsed_report);
+...
+```
+
+### 2. Challenger is provided with collaterals:
+In this scenario the attester provides the evidence and collaterals to the challenger.  The challenger then uses these to verify the TEE.
+
+##### Attester generates the evidence and collaterals (inside an enclave/TEE)
+```C
+...
+result = oe_get_report(
+    OE_REPORT_FLAGS_REMOTE_ATTESTATION,
+    NULL, // report_data
+    0,
+    NULL, // opt_params
+    0,
+    &report,
+    report_size);
+
+result = oe_get_collaterals(
```

Comment:
We will need differentiate between collaterals that are from OE SDK and collaterals from a different source.  Will address this later.

Reply:
As part of the verification, the identity of the enclave (in SGX this is the QE identity info) in the endorsement is used to verify the evidence. This already a check that is being done as part of verification.  Furthermore, in the SGX case you need to get the quote/evidence before you can get the collaterals/endorsements because the CRL depend on the PCK Cert that is embedded in the quote.